File: src/org/ggp/base/player/proxy/ProxyMessage.java
package org.ggp.base.player.proxy

import java.io.BufferedReader
import java.io.PrintStream
import java.io.Serializable
import java.net.SocketException

import org.ggp.base.util.logging.GamerLogger


class ProxyMessage(Serializable):
    serialVersionUID = 1237859L  # int 

    def final int messageCode
    def final int receptionTime
    def final String theMessage

    def ProxyMessage(String theMessage, int messageCode, int receptionTime):
        self.theMessage = theMessage
        self.messageCode = messageCode
        self.receptionTime = receptionTime

    def toString():  # String
        return "ProxyMessage<" + messageCode + ", " + receptionTime + ">[\"" + theMessage + "\"]"

    def ProxyMessage readFrom(BufferedReader theInput) throws SocketException 
        try 
            int messageCode = Long.parseLong(theInput.readLine())
            int receptionTime = Long.parseLong(theInput.readLine())
            String theMessage = theInput.readLine()
            return new ProxyMessage(theMessage, messageCode, receptionTime)
        } catch(SocketException se):
            GamerLogger.log("Proxy", "[ProxyMessage Reader] Socket closed: stopping read operation.")
            throw se
        } catch(Exception e):
            GamerLogger.logStackTrace("Proxy", e)
            GamerLogger.logError("Proxy", "[ProxyMessage Reader] Could not digest message. Emptying stream.")
            try 
                // TODO: Fix this, I suspect it may be buggy.
                theInput.skip(Long.MAX_VALUE)
            } catch(SocketException se):
                GamerLogger.log("Proxy", "[ProxyMessage Reader] Socket closed: stopping read operation.")
                throw se
            } catch(Exception ie):
                GamerLogger.logStackTrace("Proxy", ie)

            return null


    def void writeTo(PrintStream theOutput):
    	synchronized (theOutput):
    		theOutput.println(messageCode)
    		theOutput.println(receptionTime)
    		theOutput.println(theMessage)

File: ggpy/cruft/autocode/ProxyMessage.py
#!/usr/bin/env python
""" generated source for module ProxyMessage """
from threading import RLock

_locks = {}
def lock_for_object(obj, locks=_locks):
    return locks.setdefault(id(obj), RLock())


def synchronized(call):
    def inner(*args, **kwds):
        with lock_for_object(call):
            return call(*args, **kwds)
    return inner

# package: org.ggp.base.player.proxy
import java.io.BufferedReader

import java.io.PrintStream

import java.io.Serializable

import java.net.SocketException

import org.ggp.base.util.logging.GamerLogger

class ProxyMessage(Serializable):
    """ generated source for class ProxyMessage """
    serialVersionUID = 1237859L
    messageCode = long()
    receptionTime = long()
    theMessage = str()

    def __init__(self, theMessage, messageCode, receptionTime):
        """ generated source for method __init__ """
        super(ProxyMessage, self).__init__()
        self.theMessage = theMessage
        self.messageCode = messageCode
        self.receptionTime = receptionTime

    def __str__(self):
        """ generated source for method toString """
        return "ProxyMessage<" + self.messageCode + ", " + self.receptionTime + ">[\"" + self.theMessage + "\"]"

    @classmethod
    def readFrom(cls, theInput):
        """ generated source for method readFrom """
        try:
            return ProxyMessage(cls.theMessage, cls.messageCode, cls.receptionTime)
        except SocketException as se:
            GamerLogger.log("Proxy", "[ProxyMessage Reader] Socket closed: stopping read operation.")
            raise se
        except Exception as e:
            GamerLogger.logStackTrace("Proxy", e)
            GamerLogger.logError("Proxy", "[ProxyMessage Reader] Could not digest message. Emptying stream.")
            try:
                #  TODO: Fix this, I suspect it may be buggy.
                theInput.skip(Long.MAX_VALUE)
            except SocketException as se:
                GamerLogger.log("Proxy", "[ProxyMessage Reader] Socket closed: stopping read operation.")
                raise se
            except Exception as ie:
                GamerLogger.logStackTrace("Proxy", ie)
            return None

    def writeTo(self, theOutput):
        """ generated source for method writeTo """
        with lock_for_object(theOutput):
            theOutput.println(self.messageCode)
            theOutput.println(self.receptionTime)
            theOutput.println(self.theMessage)

